webapp/lib/controller.dart

```diff
@@ -108,71 +119,117 @@ void command(UIAction action, Data actionData) {
-        key: (item) => (item as model.NeedsReplyData).datetime,
-        value: (item) => (item as model.NeedsReplyData).needsReplyCount);
-      view.contentView.needsReplyTimeseries.updateChart([data]);
-
-      data = new Map.fromIterable(selectedProjectNeedsReplyDataList,
-        key: (item) => (item as model.NeedsReplyData).datetime,
-        value: (item) => (item as model.NeedsReplyData).needsReplyAndEscalateCount);
-      view.contentView.needsReplyAndEscalateTimeseries.updateChart([data]);
-
-
-      data = new Map.fromIterable(selectedProjectNeedsReplyDataList,
-        key: (item) => (item as model.NeedsReplyData).datetime,
-        value: (item) => (item as model.NeedsReplyData).needsReplyMoreThan24h);
-      view.contentView.needsReplyMoreThan24hTimeseries.updateChart([data]);
+      List<model.NeedsReplyData> selectedProjectNeedsReplyDataList = [];
 
+      if (actionData !=null && actionData is ChartFilterdata) {
+        DateTime filterDate = getFilteredDate(actionData);
+        selectedProjectNeedsReplyDataList = needsReplyDataList.where((d) => 
+            d.project == view.contentView.projectSelectorView.selectedProject &&
+            d.datetime.isAfter(filterDate)).toList();
+      } else {
+        selectedProjectNeedsReplyDataList = needsReplyDataList.where((d) => 
+            d.project == view.contentView.projectSelectorView.selectedProject).toList();
+      }
 
-      data = new Map.fromIterable(selectedProjectNeedsReplyDataList,
-        key: (item) => (item as model.NeedsReplyData).datetime,
-        value: (item) => (item as model.NeedsReplyData).needsReplyAndEscalateMoreThan24hCount);
-      view.contentView.needsReplyAndEscalateMoreThan24hTimeseries.updateChart([data]);
-
-
-      DateTime latestDateTime = data.keys.reduce((dt1, dt2) => dt1.isAfter(dt2) ? dt1 : dt2);
-      var latestData = selectedProjectNeedsReplyDataList.firstWhere((d) => d.datetime == latestDateTime, orElse: () => null);
-
-      view.contentView.needsReplyLatestValue.updateChart('${latestData.needsReplyCount}');
-      view.contentView.needsReplyAndEscalateLatestValue.updateChart('${latestData.needsReplyAndEscalateCount}');
-      view.contentView.needsReplyMoreThan24hLatestValue.updateChart('${latestData.needsReplyMoreThan24h}');
-      view.contentView.needsReplyAndEscalateMoreThan24hLatestValue.updateChart('${latestData.needsReplyAndEscalateMoreThan24hCount}');
-
-      view.contentView.needsReplyAgeHistogram.updateChart(latestData.needsReplyMessagesByDate);
+      updateNeedsReplyCharts(selectedProjectNeedsReplyDataList);
       break;
 
     case UIAction.systemEventsDataUpdated:
-      if (actionData != null && (actionData as ProjectSelectionData).isProjectSelection) {
+      if (actionData != null && actionData is ProjectSelectionData &&  actionData.isProjectSelection) {
         view.contentView.populateUrlFilters();
       } else {
         view.contentView.changeViewOnUrlChange();
       }
```

Comment:
This is getting a bit unwieldy with multiple types of data objects being passed to the same command. I'd rather us have one type of data for each command. Perhaps you can introduce some actions for the project selection change and for the date filter change, and for anything else that's needed? It should also clear up the logic quite a bit, at the moment is a bit difficult to follow. Something for a follow up PR I think?

Reply:
I agree. Created new action for chart filtering. I'll do the same for project selection in a separate PR

Reply:
great, thanks!